ASTScope: Redo 'selfDC' computation

slavapestov opened the pull request:
This was motivated by getting `self` in capture lists working without parse-time name lookup.

slavapestov committed:
ASTScope: Remove unused shouldCreateAccessorScope() method

slavapestov committed:
AST: Capture list bindings now point back to their parent CaptureListExpr

We'll need this to get the right 'selfDC' when name lookup
finds a 'self' declaration in a capture list, eg

class C {
  func bar() {}
  func foo() {
    _ = { [self] in bar() }
  }
}

slavapestov committed:
AST: Generalize CaptureListEntry::isSimpleSelfCapture() to work without parse-time lookup

The DeclRefExpr here was resolved by parse-time lookup.
Without it, it's an UnresolvedDeclRefExpr.

slavapestov committed:
ASTScope: Compute 'selfDC' directly in ASTScopeDeclConsumerForUnqualifiedLookup

Instead of having ASTScope compute 'selfDC', the lookup
consumer can compute it on its own, by looking for
bindings named 'self'.

slavapestov committed:
ASTScope: Remove old implementation of 'selfDC' computation

slavapestov commented:
https://github.com/apple/llvm-project/pull/1820
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

davidungar commented on lib/AST/Expr.cpp:
```diff
@@ -1243,7 +1251,12 @@ CaptureListExpr *CaptureListExpr::create(ASTContext &ctx,
   auto size = totalSizeToAlloc<CaptureListEntry>(captureList.size());
   auto mem = ctx.Allocate(size, alignof(CaptureListExpr));
-  return ::new(mem) CaptureListExpr(captureList, closureBody);
+  auto *expr = ::new(mem) CaptureListExpr(captureList, closureBody);
+
+  for (auto capture : captureList)
+    capture.Var->setParentCaptureList(expr);
+
+  return expr;
```
I would prefer putting the for loop *before* the `CaptureListExpr` constructor to compute the nullable `CaptureListExpr`, then passing it in to the constructor. That way the instance variable could be a `const` and it would be a bit clearer when reading the code that it is immutable.

slavapestov replied:
I'm not sure I understand. The for loop references the newly-constructed CaptureListExpr instance.

davidungar replied:
Sorry, my mistake. Thanks for pointing it out so gently.

davidungar replied:
I guess the real issue is not knowing where the var is captured when the var is created. Would be great if the var could be immutable somehow.

davidungar commented on include/swift/AST/Decl.h:
```diff
@@ -5091,8 +5092,22 @@ class VarDecl : public AbstractStorageDecl {
+  /// var decl.
+  void setParentCaptureList(CaptureListExpr *expr) {
+    assert(expr != nullptr);
+    Parent = expr;
```
Can `Parent` be a `const` or do `VarDecls` move from family to family?

davidungar commented on include/swift/AST/Decl.h:
```diff
@@ -352,13 +353,10 @@ class alignas(1 << DeclAlignInBits) Decl {
-    /// Whether this declaration was an element of a capture list.
-    IsCaptureList : 1,
-                        
     /// Whether this declaration captures the 'self' param under the same name.
```
Nice clean up.

davidungar commented on lib/AST/ASTScopeLookup.cpp:
```diff
@@ -462,69 +461,6 @@ bool ASTScopeImpl::lookupLocalBindingsInPattern(const Pattern *p,
-
-NullablePtr<DeclContext>
-GenericTypeOrExtensionWhereOrBodyPortion::checkNestedScopesForSelfCapture(
-    ArrayRef<const ASTScopeImpl *> history, size_t start) {
-  NullablePtr<DeclContext> innerCapturedSelfDC;
-  // Start with the next scope down the tree.
-  size_t j = start;
-
-  // Note: even though having this loop nested inside the while loop from
-  // GenericTypeOrExtensionWhereOrBodyPortion::computeSelfDC may appear to
-  // result in quadratic blowup, complexity actually remains linear with respect
-  // to the size of history. This relies on the fact that
-  // GenericTypeOrExtensionScope::computeSelfDCForParent returns a null pointer,
-  // which will cause this method to bail out of the search early. Thus, this
-  // method is called once per type body in the lookup history, and will not
-  // end up re-checking the bodies of nested types that have already been
-  // covered by earlier calls, so the total impact of this method across all
-  // calls in a single lookup is O(n).
-  while (j != 0) {
-      auto *entry = history[--j];
-    Optional<NullablePtr<DeclContext>> selfDCForParent =
-      entry->computeSelfDCForParent();
-
-    // If we encounter a scope that should cause us to forget the self
-    // context (such as a nested type), bail out and use whatever the
-    // the last inner captured context was.
-    if (selfDCForParent && (*selfDCForParent).isNull())
-      break;
-
-    // Otherwise, if we have a captured self context for this scope, then
-    // remember it since it is now the innermost scope we have encountered.
-    NullablePtr<DeclContext> capturedSelfDC = entry->capturedSelfDC();
-    if (!capturedSelfDC.isNull())
-      innerCapturedSelfDC = entry->capturedSelfDC();
-
-    // Continue searching in the next scope down.
-  }
-  return innerCapturedSelfDC;
-}
-
```
Lovely!

davidungar commented on lib/AST/ASTScopeLookup.cpp:
```diff
@@ -593,63 +529,6 @@ NullablePtr<const ASTScopeImpl> ASTScopeImpl::ancestorWithDeclSatisfying(
-}
-
-Optional<NullablePtr<DeclContext>>
-PatternEntryInitializerScope::computeSelfDCForParent() const {
-  // Pattern binding initializers are only interesting insofar as they
-  // affect lookup in an enclosing nominal type or extension thereof.
-  if (auto *ic = getPatternEntry().getInitContext()) {
-    if (auto *bindingInit = dyn_cast<PatternBindingInitializer>(ic)) {
-      // Lazy variable initializer contexts have a 'self' parameter for
-      // instance member lookup.
-      if (bindingInit->getImplicitSelfDecl()) {
-        return NullablePtr<DeclContext>(bindingInit);
-      }
-    }
-  }
-  return None;
-}
-
-Optional<NullablePtr<DeclContext>>
-MethodBodyScope::computeSelfDCForParent() const {
-  return NullablePtr<DeclContext>(decl);
-}
-
-#pragma mark capturedSelfDC
-
-// Closures may explicitly capture the self param, in which case the lookup
-// should use the closure as the context for implicit self lookups.
-
-// By default, there is no such context to return.
-NullablePtr<DeclContext> ASTScopeImpl::capturedSelfDC() const {
-  return NullablePtr<DeclContext>();
-}
-
-// Closures track this information explicitly.
-NullablePtr<DeclContext> ClosureParametersScope::capturedSelfDC() const {
-  if (closureExpr->capturesSelfEnablingImplictSelf())
-    return NullablePtr<DeclContext>(closureExpr);
-  return NullablePtr<DeclContext>();
-}
-
```
Nice

davidungar commented on lib/AST/UnqualifiedLookup.cpp:
```diff
@@ -553,6 +557,49 @@ bool ASTScopeDeclConsumerForUnqualifiedLookup::consume(
+    if (auto *var = dyn_cast<VarDecl>(value)) {
+      // We have a binding named 'self'.
+      //
+      // There are three possibilities:
+      //
+      // 1) This binding is the 'self' parameter of a method,
+      // 2) This binding is a bona-fide 'self' capture, meaning a capture
+      //    list entry named 'self' with initial value expression 'self',
+      // 3) None of the above.
+      if (var->getName() == factory.Ctx.Id_self) {
+        // How we handle these cases depends on whether we've already seen
+        // another 'self' binding.
+        if (candidateSelfDC == nullptr) {
+          // We haven't seen one yet, so record it.
+          if (var->isSelfParameter())
+            candidateSelfDC = var->getDeclContext();
+          else if (var->isSelfParamCapture())
+            candidateSelfDC = var->getParentCaptureList()->getClosureBody();
+        } else {
+          // If we see a binding named 'self' that is not a bona-fide
+          // 'self', we have to forget about the previous 'self' capture
+          // because it's not going to be the right one for accessing
+          // instance members of the innermost nominal type. Eg,
+          //
+          // class C {
+          //   func bar() {}
+          //   func foo() {
+          //     _ { [self=12] { [self] bar() } }
+          //   }
+          // }
+          //
+          // Instead, we're going to move on and look for the next-innermost
+          // 'self' binding.
+          if (!var->isSelfParameter() &&
+              !var->isSelfParamCapture())
+            candidateSelfDC = nullptr;
+        }
+      }
+    }
+
```
The comments are great! But, at this point, this function is too long for me to process, even though most of it is comments. I suspect you could break it up into very manageable pieces. Maybe split off the body of the for loop into one piece, and split the two halves of the `if` on 563? That's just what I would try. You can find something better, I bet. Something that communicates what's really going on here, succinctly at the upper levels.

slavapestov replied:
I split off the body of the `if` statement into a new `maybeUpdateSelfDC()` method.

davidungar commented on lib/AST/UnqualifiedLookup.cpp:
```diff
@@ -586,22 +633,36 @@ bool ASTScopeDeclGatherer::consume(ArrayRef<ValueDecl *> valuesArg,
     }
   }
+
+  // We're looking for members of a type.
+  //
+  // If we started the looking from inside a scope where a 'self' parameter
+  // is visible, instance members are returned with the 'self' parameter's
+  // DeclContext as the base, which is how the expression checker knows to
+  // convert the unqualified reference into a self member access.
   auto resultFinder = UnqualifiedLookupFactory::ResultFinderForTypeContext(
-      &factory, selfDC ? selfDC.get() : scopeDC, scopeDC);
+      &factory, candidateSelfDC ? candidateSelfDC : scopeDC, scopeDC);
   const bool isCascadingUse =
       calculateIsCascadingUse(factory.getInitialIsCascadingUse());
   factory.findResultsAndSaveUnavailables(scopeDC, std::move(resultFinder),
                                          isCascadingUse, factory.baseNLOptions);
   factory.recordCompletionOfAScope();
+
+  // We're done looking inside a nominal type declaration. It is possible
+  // that this nominal type is nested inside of another type, in which case
+  // we will visit the outer type next. Make sure to clear out the known
+  // 'self' parameeter context, since any members of the outer type are
+  // not accessed via the innermost 'self' parameter.
+  candidateSelfDC = nullptr;
```
The new comments are great!

davidungar reviewed (approved):
I love this direction. Didn't look deeply enough to assure correctness, but very nice to shift the complexity as this is doing. It wouldn't be a review from me without a request to break up a long function, where I think clarity could be improved.

slavapestov committed:
ASTScope: Collapse PureFunctionBodyScope and MethodBodyScope

This centralizes some invariants around the 'self' parameter.
While all ConstructorDecls and DestructorDecls have a 'self',
even if they're invalid because they're not nested inside a type,
we don't want to consider this as the base 'self' for lookups.

Eg, consider this invalid code:

class C {
  func f() {
    init() {
      x
    }
  }
}

The base for the lookup should be f.self, not f.init.self.

slavapestov committed:
AST: Remove VarDecl::hasNonPatternBindingInit()

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

davidungar commented on lib/AST/UnqualifiedLookup.cpp:
```diff
@@ -547,12 +553,61 @@ void UnqualifiedLookupFactory::lookInASTScopes() {
+    // references. This is used by lookInMembers().
+    if (auto *var = dyn_cast<VarDecl>(value)) {
+      if (var->getName() == factory.Ctx.Id_self) {
+        maybeUpdateSelfDC(var);
```
Now that you've refactored--thanks!--I can see more clearly what is going on with candidateSelfDC. Is it a fragile piece of mutable state? Could we pass a piece of immutable state through the lookup recursion instead? It seems to be a bit odd that a `consume` method has a side effect. What do you think?

slavapestov replied:
I was actually going to try to refactor the lookup to use a loop that walks the parents instead of recursing. This will allow us to implement the one-level lookup we need to do re-declaration checking for locals defined in the same scope. Maybe then the selfDC computation can go there and be totally explicit, instead of being part of consume().

davidungar reviewed (approved):
The `selfDC` computation is better, but maybe in a future PR it could be even better.

slavapestov commented:
apple/llvm-project#1820
@swift-ci Please smoke test

slavapestov commented:
apple/llvm-project#1820
@swift-ci Please smoke test

slavapestov commented:
apple/llvm-project#1820
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
https://github.com/apple/llvm-project/pull/1820
@slavapestov Please smoke test

slavapestov commented:
apple/llvm-project#1820
@slavapestov Please smoke test

slavapestov commented:
apple/llvm-project#1820
@slavapestov Please smoke test

slavapestov commented:
apple/llvm-project#1820
@slavapestov Please smoke test

slavapestov commented:
@slavapestov Please test source compatibility

slavapestov commented:
@slavapestov Please test source compatibility

slavapestov commented:
@slavapestov Please test source compatibility

slavapestov commented:
apple/llvm-project#1820
@slavapestov Please smoke test

slavapestov commented:
apple/llvm-project#1820
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility
